Introduce native arch support in arch database

mirza-halilcevic opened the pull request:
- Resolves https://github.com/ROCm/rocMLIR-internal/issues/1745

mirza-halilcevic committed:
Introduce native arch support in arch database.

Signed-off-by: Mirza Halilcevic <mirza.halilcevic@amd.com>

dhernandez0 commented on mlir/lib/Dialect/Rock/utility/CMakeLists.txt:
```diff
@@ -19,4 +19,14 @@ add_rocmlir_dialect_library(MLIRRockUtility
+)
+target_link_libraries(obj.MLIRRockUtility PUBLIC
+  hip::host hip::amdhip64
 )
```
new line

dhernandez0 commented on mlir/unittests/Dialect/Rock/AmdArchDbTests.cpp:
```diff
@@ -0,0 +1,20 @@
+//===- TransformMapBuilderTests.cpp - Tests for the Rock Transform
```
update comment

dhernandez0 commented on mlir/unittests/Dialect/Rock/AmdArchDbTests.cpp:
```diff
@@ -0,0 +1,20 @@
+
+TEST(AmdArchDbTest, NativeArch) {
+  auto info = lookupArchInfo("native");
+  EXPECT_EQ(info.totalVGPRPerEU, 256);
```
the current arch in the system could have a different number, right?

mirza-halilcevic replied:
Yes, of course. I just used this test as a convenient way to experiment. This will be changed later.

dhernandez0 commented:
> * Resolves [Improve gpu attribute collection technique rocMLIR-internal#1745](https://github.com/ROCm/rocMLIR-internal/issues/1745)

I think apart from the "native" feature. We need to have a python version of AmdArchDb, because right now we hardcode lots of hardware specific details in our python codebase. So, I guess let's create a ticket for that and reduce the scope of https://github.com/ROCm/rocMLIR-internal/issues/1745 to be only for C++.

dhernandez0 commented on mlir/lib/Dialect/Rock/utility/AmdArchDb.cpp:
```diff
@@ -73,11 +75,36 @@ static constexpr AmdArchInfo
+  // are these the correct values to use?
+  ret.totalSharedMemPerCU = prop.sharedMemPerMultiprocessor;
+  ret.maxSharedMemPerWG = prop.sharedMemPerBlock;
+
```
is there a way to get the number of CUs? that can change even with the same arch id (gfx90a could be MI210 and MI250, and numCU is different)

mirza-halilcevic replied:
Yes, there is `prop.multiProcessorCount`. One thing to note though is that `prop.multiProcessorCount` on _gfx1200_ shows half the number `rocminfo` shows. For example, if `rocminfo` shows 32 CUs, `prop.multiProcessorCount` will be 16. So we might need to multiply that value by 2, depending on what the caller expects. I still need to confirm if the behavior is the same on other architectures.

dhernandez0 replied:
that only applies to navi because of WGP Mode I think? @umangyadav

umangyadav replied:
Yes by default we compile Navi machines for Workgroup processor mode which counts 2 CUs == 1CUs. 

That is why @mirza-halilcevic  you see half the number of CUs. We should use half the number of CUs for the Navi in archdb compared to what rocminfo shows.

dhernandez0 commented on mlir/lib/Dialect/Rock/utility/AmdArchDb.cpp:
```diff
@@ -73,11 +75,36 @@ static constexpr AmdArchInfo
+    if (!llvm::getAsUnsignedInteger(remainingParts, 0, deviceId)) {
+      return fetchNativeArchInfo(deviceId);
+    }
+    return fetchNativeArchInfo();
```
I'd prefer to fail if no gpu id is provided. What do you guys think?

mirza-halilcevic replied:
From what I saw, CUDA and HIP generally default to device 0 if not specified otherwise. That's the convention I followed here.

But either way is fine with me if you want me to change this. I guess it depends on how often we expect to encounter non-homogenous systems, and how much of a hassle it is for the caller to attach a device ID to the argument.

umangyadav replied:
>From what I saw, CUDA and HIP generally default to device 0 if not specified otherwise. That's the convention I followed here. 

I agree with keeping this convention

dhernandez0 replied:
Sounds reasonable although I prefer the "keep it simple" principle (we can require migraphx to provide the gpu id). If you guys prefer to default to 0, let's do it.

umangyadav commented on mlir/lib/Dialect/Rock/utility/AmdArchDb.cpp:
```diff
@@ -73,11 +75,36 @@ static constexpr AmdArchInfo
               /*hasFp8ConversionInstrs=*/false,
               /*hasOcpFp8ConversionInstrs=*/false, /*maxNumXCC=*/1);
 
+static AmdArchInfo fetchNativeArchInfo(unsigned long long deviceId = 0) {
```
nit : you don't need to use  very wide dtype for `deviceId`

umangyadav commented on mlir/lib/Dialect/Rock/utility/AmdArchDb.cpp:
```diff
@@ -73,11 +75,36 @@ static constexpr AmdArchInfo
+
+  auto ret = lookupArchInfo(prop.gcnArchName); // get baseline
+  ret.waveSize = prop.warpSize;
+  // are these the correct values to use?
```
Yes they look correct to me.

umangyadav commented on mlir/lib/Dialect/Rock/utility/AmdArchDb.cpp:
```diff
@@ -73,11 +75,36 @@ static constexpr AmdArchInfo
+  }
+
+  auto ret = lookupArchInfo(prop.gcnArchName); // get baseline
+  ret.waveSize = prop.warpSize;
```
Can you please check what warpSize are you getting for the Navi ?

mirza-halilcevic replied:
On Navi4x, I get a warp size of 32.

umangyadav commented on mlir/lib/Dialect/Rock/utility/CMakeLists.txt:
```diff
@@ -19,4 +19,14 @@ add_rocmlir_dialect_library(MLIRRockUtility
+find_package(hip REQUIRED)
+# We do need both of these.  One is for linking, to get the HIP
+# functions.  The other is for compiling, to get -D__HIP_PLATFORM_AMD__
+target_link_libraries(MLIRRockUtility PUBLIC
+  hip::host hip::amdhip64
+)
+target_link_libraries(obj.MLIRRockUtility PUBLIC
+  hip::host hip::amdhip64
```
This is strange that you need to use `target_link_libraries` twice. I think even if you use target_link_libraries with `MLIRRockUtility` it should work

mirza-halilcevic replied:
I copied this from [here](https://github.com/ROCm/rocMLIR/blob/b6c726a324d2807494f770ed915af0e54cc6cb3f/external/llvm-project/mlir/lib/ExecutionEngine/CMakeLists.txt#L414). If I leave out the second one, I get compilation errors because `__HIP_PLATFORM_AMD__` isn't defined.

umangyadav replied:
you can try doing 
`target_compile_definitions(MLIRRockUtility PRIVATE __HIP_PLATFORM_AMD__)`

umangyadav replied:
https://github.com/ROCm/rocm-examples/blob/f9d4e5e78325c36b319d91ec37c6410b2b6e12fb/HIP-Basic/device_query/CMakeLists.txt#L83

umangyadav replied:
> you can try doing `target_compile_definitions(MLIRRockUtility PRIVATE __HIP_PLATFORM_AMD__)`


@mirza-halilcevic  did you try this one ?

mirza-halilcevic replied:
Yes, I tried. But unfortunately, I couldn't get it to work with that approach.

umangyadav commented:
> So, I guess let's create a ticket for that and reduce the scope of https://github.com/ROCm/rocMLIR-internal/issues/1745 to be only for C++. 

I wonder if we can use pybind or some other mechanism available in upstream llvm to use C++ API from Python

mirza-halilcevic committed:
- Expand the native arch info fetch with LLVM base info values.
- Implement python bindings.

mirza-halilcevic committed:
Move gfx12Info to a global constant.

dhernandez0 commented on mlir/lib/Dialect/Rock/utility/AmdArchDb.cpp:
```diff
@@ -71,18 +78,129 @@ static constexpr AmdArchInfo
               /*hasFp8ConversionInstrs=*/false,
-              /*hasOcpFp8ConversionInstrs=*/false, /*maxNumXCC=*/1);
+              /*hasOcpFp8ConversionInstrs=*/false, /*maxNumXCC=*/1),
+    gfx12Info(GemmFeatures::dot | GemmFeatures::atomic_add |
```
rdna4Info?

mirza-halilcevic replied:
Sure. In that case we should also change `gfx11Info` to `rdna3Info`. And regarding `gfx950Info` in https://github.com/ROCm/rocMLIR/pull/1802, maybe we can change that to `cdna3_5Info` or something similar?

dhernandez0 commented on mlir/lib/Dialect/Rock/utility/AmdArchDb.cpp:
```diff
@@ -71,18 +78,129 @@ static constexpr AmdArchInfo
+                  GemmFeatures::atomic_fmax_f32 | GemmFeatures::wmma |
+                  GemmFeatures::atomic_add_f16 | GemmFeatures::atomic_add_bf16,
+              /*waveSize=*/32, /*maxWavesPerEU*/ 16, /*totalSGPRPerEU*/ 800,
+              /*totalVGPRPerEU*/ 1536, /*totalSharedMemPerCU*/ 131072,
```
totalSharedMemPerCU was 65536 in the existing code. Is this a mistake or is 131072 the correct value?

mirza-halilcevic replied:
I believe that the previous value was incorrect. If we look at the others, we can see that for RDNA, the `totalSharedMemPerCU` is two times the `maxSharedMemPerWG`.

LLVM gives us the following values:
```
AddressableLocalMemorySize: 65536
LocalMemorySize: 131072
```
It seems like the `LocalMemorySize` should correspond to `totalSharedMemPerCU`.

HIP returns the following:
```
MaxSharedMemoryPerBlock: 65536
MaxSharedMemoryPerMultiprocessor: 65536
SharedMemPerMultiprocessor: 1048576 // NOTE: No idea what this value represents
```
That's why the initial mistake was made.

dhernandez0 replied:
yes, that sounds right based on the comments here (we use WGP mode for navi): https://github.com/llvm/llvm-project/blob/f04bfbc4162247c0074fdd87068053e4264dbefa/llvm/lib/Target/AMDGPU/AMDGPUSubtarget.h#L269

Thanks!

dhernandez0 commented on mlir/lib/Dialect/Rock/utility/AmdArchDb.cpp:
```diff
@@ -71,18 +78,129 @@ static constexpr AmdArchInfo
+    LLVMInitializeAMDGPUTargetMC();
+  });
+
+  Triple triple("amdgcn-amd-");
```
shouldn't this be amdgcn-amd-amdhsa?

mirza-halilcevic replied:
I used [this](https://github.com/llvm/llvm-project/blob/209252f3d5bf66c8bff0863ba7617e37ead4fff8/llvm/unittests/Target/AMDGPU/AMDGPUUnitTests.cpp#L173) as a reference. It seems to be what LLVM is expecting.

dhernandez0 replied:
I think our code always generates amdgcn-amd-amdhsa (you can see that in our tests if you search "amdgcn-amd-". I'm not entirely sure of the differences between them: https://llvm.org/docs/AMDGPUUsage.html#amdgpu-target-triples

mirza-halilcevic replied:
Looks like both work fine. I set it to `amdgcn-amd-amdhsa`.

dhernandez0 commented on mlir/lib/Dialect/Rock/utility/AmdArchDb.cpp:
```diff
@@ -71,18 +78,129 @@ static constexpr AmdArchInfo
+}
+
+std::unique_ptr<const GCNTargetMachine>
+createTargetMachine(StringRef cpu, StringRef featureString = "") {
```
cpu -> chip? as this won't be a cpu

mirza-halilcevic replied:
CPU is the terminology that the LLVM Target related identifiers use, regardless of the actual architecture. That's why I kept it like this, but it's no trouble to change it if you think it's semantically more appropriate.

dhernandez0 commented on mlir/lib/Dialect/Rock/utility/AmdArchDb.cpp:
```diff
@@ -71,18 +78,129 @@ static constexpr AmdArchInfo
+  TargetOptions options;
+  return std::unique_ptr<GCNTargetMachine>(
+      static_cast<GCNTargetMachine *>(target->createTargetMachine(
+          triple, cpu, featureString, options, std::nullopt, std::nullopt)));
```
nit: "options, std::nullopt, std::nullopt" -> "{}, {}" (as in ModuleToObject.cpp)

dhernandez0 commented on mlir/lib/Dialect/Rock/utility/AmdArchDb.cpp:
```diff
@@ -71,18 +78,129 @@ static constexpr AmdArchInfo
+  }
+}
+
+AmdArchInfo fetchNativeArchInfo(unsigned deviceId = 0) {
```
I guess ROCR_VISIBLE_DEVICES / HIP_VISIBLE_DEVICES etc filter out gpus so that gpu 0 is already what we want, right?

mirza-halilcevic replied:
Yes. CUDA has the same behavior. The device IDs get remapped so that from the programmer's point of view, the device ID is always `[0, N-1]`.

dhernandez0 commented on mlir/lib/Dialect/Rock/utility/AmdArchDb.cpp:
```diff
@@ -71,18 +78,129 @@ static constexpr AmdArchInfo
+  if (auto err = hipGetDeviceProperties(&prop, deviceId); err != hipSuccess) {
+    llvm::errs() << "WARNING: hipGetDeviceProperties failed with error: "
+                 << hipGetErrorString(err) << ". Falling back to defaults.\n";
+    return gcnInfo;
```
The default for rdna should be rdnaNoDotInfo. Otherwise, the waveSize would be wrong. I guess we don't know if it's navi or MI at this point... Not sure what is the best to do. Maybe just fail?

mirza-halilcevic replied:
That's a good point. If you look at the `lookupArchInfo` implementation, it also returns `gcnInfo` as a fallback always. I would agree that it's probably best to just fail.

dhernandez0 replied:
let's fail then? @umangyadav

dhernandez0 commented on mlir/lib/Dialect/Rock/utility/AmdArchDb.cpp:
```diff
@@ -71,18 +78,129 @@ static constexpr AmdArchInfo
+  checkAndSetInfo("(LLVM) totalSharedMemPerCU", ret.totalSharedMemPerCU,
+                  st.getAddressableLocalMemorySize());
+  checkAndSetInfo("(LLVM) maxSharedMemPerWG", ret.maxSharedMemPerWG,
+                  st.getLocalMemorySize());
```
I guess the only missing things are:
- fp8/ocp -> you can get that with amdgpu::hasOcpFp8(chipset), see rocmlir-gen.cpp
- maxNumXCC
- GemmFeatures: mfma/wmma, dot, atomic_add, atomic_add_f16, atomic_add_bf16, atomic_fmax_f32

Any chance we can get those as well from LLVM?

mirza-halilcevic replied:
I'll take a look if we can fetch those too.

dhernandez0 replied:
yes, that sounds right based on the comments here: https://github.com/llvm/llvm-project/blob/f04bfbc4162247c0074fdd87068053e4264dbefa/llvm/lib/Target/AMDGPU/AMDGPUSubtarget.h#L269

we use WGP mode for navi

dhernandez0 replied:
for atomic, there are:
- hasAtomicBufferPkAddBF16Inst: we use buffer atomics only for now for f16/bf16
- HasAtomicBufferGlobalPkAddF16NoRtnInsts and HasAtomicBufferGlobalPkAddF16Insts 
- HasAtomicFMinFMaxF32GlobalInsts (I guess it's this one for atomic_fmax_f32)
- hasAtomicFaddInsts (atomic_add)

dhernandez0 commented on mlir/lib/Dialect/Rock/utility/Bindings/AmdArchDbBindings.cpp:
```diff
@@ -0,0 +1,55 @@
+PYBIND11_MODULE(amd_arch_db, m) {
+  m.doc() = "Database of AMD GPU features";
+
+  py::enum_<mlir::rock::GemmFeatures>(m, "GemmFeatures")
```
add comment in GemmFeatures definition to keep it in sync with python bindings

dhernandez0 commented on mlir/lib/Dialect/Rock/utility/Bindings/AmdArchDbBindings.cpp:
```diff
@@ -0,0 +1,55 @@
+      .value("ATOMIC_ADD_F16", mlir::rock::GemmFeatures::atomic_add_f16)
+      .value("ATOMIC_FMAX_F32", mlir::rock::GemmFeatures::atomic_fmax_f32);
+
+  py::class_<mlir::rock::AmdArchInfo>(m, "AmdArchInfo")
```
add comment in AmdArchInfo definition to keep it in sync with python bindings

dhernandez0 commented on mlir/lib/Dialect/Rock/utility/Bindings/AmdArchDbBindings.cpp:
nice! thanks, this will improve the python code. Will you change the python code to use this here or in a follow-up PR?

mirza-halilcevic replied:
I'm leaning towards doing it in a separate PR. What would you suggest?

dhernandez0 replied:
yes, separate PR is a good idea

mirza-halilcevic committed:
Add fetching of atomics and fp8 support info, as well as address code review
comments.

mirza-halilcevic committed:
Add pybind11 as a pip requirement.

umangyadav commented:
@mirza-halilcevic  Can you resolve merge conflicts ?

dhernandez0 commented on mlir/lib/Dialect/Rock/utility/AmdArchDb.cpp:
```diff
@@ -73,24 +80,183 @@ static constexpr AmdArchInfo
+
+    checkAndSetInfo(
+        "(LLVM) hasFp8ConversionInstrs", ret.hasFp8ConversionInstrs,
+        st.hasFP8ConversionInsts()); // TODO double check the meaning of this
```
It looks like this: st.hasFP8ConversionInsts() tells us if it has fp8+bf8 support or not. 
So, if st.hasFP8ConversionInsts() == false:
- our hasFp8ConversionInstrs=false and our hasOcpFp8ConversionInstrs=false

if st.hasFP8ConversionInsts() == true:
- set our hasFp8ConversionInstrs=not amdgpu::hasOcpFp8(maybeChipset.value())
- set our hasOcpFp8ConversionInstrs=amdgpu::hasOcpFp8(maybeChipset.value())

dhernandez0 commented on mlir/lib/Dialect/Rock/utility/AmdArchDb.cpp:
```diff
@@ -113,29 +279,14 @@ AmdArchInfo mlir::rock::lookupArchInfo(StringRef arch) {
-  return gcnInfo;
+  llvm::errs() << "WARNING: Unsupported architecture: " << arch
+               << ". Falling back to defaults.\n";
+  return gcnInfo; // TODO fail instead of default?
```
I think we should fail here, what do you think @umangyadav ? (for context: https://github.com/ROCm/rocMLIR/pull/1790#discussion_r2034806314)

umangyadav replied:
yes better to fail IMO

dhernandez0 commented on mlir/lib/Dialect/Rock/utility/AmdArchDb.cpp:
```diff
@@ -73,24 +80,183 @@ static constexpr AmdArchInfo
   }
-  StringRef chip = firstPart;
 
+  // TODO check and set maxNumXCC
```
this is needed for https://github.com/ROCm/rocMLIR-internal/issues/1801 but we can skip it in this PR

dhernandez0 commented on mlir/lib/Dialect/Rock/utility/AmdArchDb.cpp:
```diff
@@ -73,24 +80,183 @@ static constexpr AmdArchInfo
+
+    auto features = ret.defaultFeatures;
+
+    if (st.hasAtomicFaddInsts()) {
```
I guess you got these from https://github.com/ROCm/rocMLIR/pull/1790#discussion_r2034908692 please check if they are correct, I didn't check. You can verify if the result makes sense for MI300, Navi4, etc...

dhernandez0 replied:
actually, as we don't have tests, instead of manually checking you can create tests that verify the values for some archs are the expected ones (gfx942, gfx1200, ...)

mirza-halilcevic committed:
Implement a diagnostic/informational test case.

mirza-halilcevic committed:
Fail instead of falling back to defaults in AmdArchDb.

mirza-halilcevic committed:
Fix fp8 features fetching logic and improve test case.

mirza-halilcevic committed:
Missing newline.

dhernandez0 commented on mlir/unittests/Dialect/Rock/AmdArchDbTests.cpp:
```diff
@@ -0,0 +1,77 @@
+  EXPECT_EQ(presetInfo.waveSize, nativeInfo.waveSize);
+  SOFT_CHECK_EQ(presetInfo.maxWavesPerEU, nativeInfo.maxWavesPerEU);
+  SOFT_CHECK_EQ(presetInfo.totalSGPRPerEU, nativeInfo.totalSGPRPerEU);
+  SOFT_CHECK_EQ(presetInfo.totalVGPRPerEU, nativeInfo.totalVGPRPerEU);
```
just out of curiosity, why can't we use EXPECT_EQ?

mirza-halilcevic replied:
Initially, the preset and native values for these parameters didn't match. I made it a soft check so the tests wouldn't fail, in case it was expected variation.

But after discussing with the team, we concluded that we shouldn't expect any variation in these values for different GPU variants of the same base arch. So, in the end, I changed it back to `EXPECT_EQ` and fixed the incorrect preset values.

mirza-halilcevic committed:
Fix incorrect values for cdna3Info, and implement workarounds for
failing test cases.

dhernandez0 commented on mlir/lib/Dialect/Rock/utility/AmdArchDb.cpp:
```diff
@@ -260,6 +263,9 @@ AmdArchInfo fetchNativeArchInfo(unsigned deviceId = 0) {
 AmdArchInfo mlir::rock::lookupArchInfo(StringRef arch) {
   // Keep this implementation in sync with
   // mlir/test/lit.site.cfg.py.in:set_arch_features()
+  if (arch.empty()) {
```
shouldn't we fail if it's empty?

mirza-halilcevic replied:
A lot of tests were failing when we introduced the `llvm_unreachable`. Most of them were invoking the function with an empty arch string. A few of them had the target triple concatenated without a delimiter, like this `amdgcn-amd-amdhsagfx942:sramecc+:xnack-`. Previously, that would work because it would return the fallback.

This is a workaround for the empty arch case. To me, going through the codebase and fixing the individual call sites of `lookupArchInfo` doesn't seem like the right approach. But let's discuss this sometime. FYI @umangyadav 

One other approach I considered is just doing a native fetch if the arch string is empty. But that caused the tests to fail with `loc("<stdin>":1:1): error: custom op 'Fetching' is unknown (tried 'builtin.Fetching' as well)`. I have yet to figure out why that is happening.

dhernandez0 replied:
I think fixing all the cases is a good approach but out of the scope of the PR. We can create a ticket to do that and remove the workarounds.

dhernandez0 replied:
fixed this in https://github.com/ROCm/rocMLIR/pull/1862, once that's merged we can fail if arch is empty or not valid.

dhernandez0 replied:
https://github.com/ROCm/rocMLIR/pull/1862 merged. Please, update the code to fail if arch is empty.

mirza-halilcevic committed:
Fix formatting.

mirza-halilcevic committed:
Fix incorrect preset info in AmdArchDb.

dhernandez0 commented on mlir/lib/Dialect/Rock/utility/AmdArchDb.cpp:
```diff
@@ -73,24 +80,196 @@ static constexpr AmdArchInfo
+    llvm::report_fatal_error(reason.c_str());
+  }
+
+  llvm::outs() << "gcnArchName: " << prop.gcnArchName << "\n";
```
llvm::outs()  -> LLVM_DEBUG(dbgs() << "mesage");

dhernandez0 replied:
you have to define: 
#define DEBUG_TYPE "rock-amd-arch-db" for example. Then, we can print out using --debug-only=rock-amd-arch-db flag

umangyadav replied:
@mirza-halilcevic  this one is still unresolved and ideally we should put all prints behind Debug symbol as daniel suggested.

umangyadav commented on mlir/lib/Dialect/Rock/utility/Bindings/AmdArchDbBindings.cpp:
```diff
@@ -0,0 +1,55 @@
+                    &mlir::rock::AmdArchInfo::hasOcpFp8ConversionInstrs)
+      .def_readonly("max_num_xcc", &mlir::rock::AmdArchInfo::maxNumXCC);
+
+  m.def("lookup_arch_info", &lookupArchInfoWrapper);
```
Why do you have to create a wrapper ?

mirza-halilcevic replied:
Becuase `lookupArchInfo` takes an `llvm::StringRef` and that causes problems with pybind11. I'll try wrapping this in a lambda instead, it'll be cleaner like that.

mirza-halilcevic committed:
Introduce caching mechanism to native arch lookup and replace outs()
with dbgs().

mirza-halilcevic committed:
Replace wrapper with lambda in python bindings.

umangyadav commented on mlir/lib/Dialect/Rock/utility/AmdArchDb.cpp:
```diff
@@ -74,24 +83,207 @@ static constexpr AmdArchInfo
+
+  LLVM_DEBUG(llvm::dbgs() << "gcnArchName: " << prop.gcnArchName << "\n");
+
+  std::lock_guard<std::mutex> lock(m);
```
nit : define mutex near it's use.

umangyadav commented on mlir/lib/Dialect/Rock/utility/Bindings/AmdArchDbBindings.cpp:
```diff
@@ -0,0 +1,53 @@
+                    &mlir::rock::AmdArchInfo::hasFp8ConversionInstrs)
+      .def_readonly("has_ocp_fp8_conversion_instrs",
+                    &mlir::rock::AmdArchInfo::hasOcpFp8ConversionInstrs)
+      .def_readonly("max_num_xcc", &mlir::rock::AmdArchInfo::maxNumXCC);
```
HIP APIs would soon allow look up for maxNumXCC. it's been implemented now in HIP APIs. Do we have tracking issue for that when we move to rocm-7.0 ?

mirza-halilcevic replied:
I created one https://github.com/ROCm/rocMLIR-internal/issues/1886

umangyadav commented on mlir/lib/Dialect/Rock/utility/AmdArchDb.cpp:
```diff
@@ -74,24 +83,207 @@ static constexpr AmdArchInfo
+  }
+}
+
+GemmFeatures &operator|=(GemmFeatures &lhs, const GemmFeatures &rhs) {
+  lhs = lhs | rhs;
+  return lhs;
+}
+
+GemmFeatures &operator&=(GemmFeatures &lhs, const GemmFeatures &rhs) {
+  lhs = lhs & rhs;
+  return lhs;
+}
```
GemmFeatures already has `bitEnumClear` and `bitEnumSet` methods that you can use

umangyadav commented on mlir/lib/Dialect/Rock/utility/AmdArchDb.cpp:
```diff
@@ -74,24 +83,207 @@ static constexpr AmdArchInfo
+    if (st.hasAtomicBufferPkAddBF16Inst()) {
+      features |= GemmFeatures::atomic_add_bf16;
+    } else {
+      features &= ~GemmFeatures::atomic_add_bf16;
```
~I am not sure if you need to explictly remove those. It would generate incorrect encoding. 
You need to remove only if they existed initially~

umangyadav commented on mlir/unittests/Dialect/Rock/AmdArchDbTests.cpp:
```diff
@@ -0,0 +1,76 @@
+
+using namespace mlir::rock;
+
+#define SOFT_CHECK_EQ(val1, val2)                                              \
```
Looks like this unused.

mirza-halilcevic committed:
Address review comments.

umangyadav commented:
@mirza-halilcevic  CI is failing, Can you fix those failures and update the branch ?

mirza-halilcevic committed:
Fix issue with finding the hip package by specifying the ROCM path.

dhernandez0 commented on mlir/lib/Dialect/Rock/utility/AmdArchDb.cpp:
please can you change cdna35Info to cdna4Info? it seems it's cdna4 https://www.amd.com/content/dam/amd/en/documents/instinct-tech-docs/instruction-set-architectures/amd-instinct-cdna4-instruction-set-architecture.pdf @umangyadav

umangyadav replied:
Yes let's change it to cdna4 then

mirza-halilcevic committed:
Rename cdna35Info to cdna4Info, and add dependency to
AMDGPUCommonTableGen.

pabloantoniom committed:
Fix test include path and replace LLVM internal files with HSA

Some arch values are not available in HIP, se we tried to fetch
them from LLVM internal files (GCNSubtarget). This however casued
build (and CI) failures.

Instead we can rely on HSA to get most of the information, thus
avoiding LLVM internal files. In other words, this now should
build succesfully.

There are still some fields (commented and with ???) that cannot
be retrieved by either HIP or HSA - we will need to investigate
what to do with those.

umangyadav commented on mlir/lib/Dialect/Rock/IR/AmdArchDb.cpp:
```diff
@@ -136,29 +136,28 @@ std::tuple<StringRef, unsigned> parseArchString(StringRef arch) {
+  //   for this).
+  int num_cpus;
+  // Output fields:
+  uint32_t simds_per_cu;
```
nit: use camelcase

umangyadav commented on mlir/lib/Dialect/Rock/IR/AmdArchDb.cpp:
```diff
@@ -168,74 +167,61 @@ AmdArchInfo fetchNativeArchInfo(const hipDeviceProp_t &prop) {
+  // checkAndSetInfo("(HSA) totalSGPRPerEU", ret.totalSGPRPerEU, ???);
+  // checkAndSetInfo("(HSA) totalVGPRPerEU", ret.totalVGPRPerEU, ???);
+  checkAndSetInfo("(HSA) waveSize", ret.waveSize, agent_info.wavefront_size);
+  // checkAndSetInfo("(HSA) defaultFeatures", ret.defaultFeatures, ???);
```
I dont' expect HSA to have target features level details as it is runtimeAPI

dhernandez0 commented on mlir/lib/Dialect/Rock/IR/AmdArchDb.cpp:
```diff
@@ -136,29 +136,28 @@ std::tuple<StringRef, unsigned> parseArchString(StringRef arch) {
+};
+
+AmdArchInfo fetchNativeArchInfo(const hipDeviceProp_t &prop,
+                                agent_info_t &agent_info) {
```
camelCase here as well

dhernandez0 commented on mlir/lib/Dialect/Rock/IR/AmdArchDb.cpp:
```diff
@@ -136,29 +136,28 @@ std::tuple<StringRef, unsigned> parseArchString(StringRef arch) {
-}
-
-AmdArchInfo fetchNativeArchInfo(const hipDeviceProp_t &prop) {
+struct agent_info_t {
```
AgentInfo (see other examples in the codebase such as struct ConvolutionDims)

dhernandez0 commented on mlir/lib/Dialect/Rock/IR/AmdArchDb.cpp:
```diff
@@ -136,29 +136,28 @@ std::tuple<StringRef, unsigned> parseArchString(StringRef arch) {
+  unsigned deviceId;
+  //   The ID of the current GPU HSA agent (weird that HSA does not have a field
+  //   for this).
+  int num_cpus;
```
this is the ID of the gpu? why is it call num_cpus?

pabloantoniom replied:
Wrong comment 😅

dhernandez0 commented on mlir/lib/Dialect/Rock/IR/AmdArchDb.cpp:
```diff
@@ -168,74 +167,61 @@ AmdArchInfo fetchNativeArchInfo(const hipDeviceProp_t &prop) {
+  checkAndSetInfo("(HSA) maxWavesPerEU", ret.maxWavesPerEU,
+                  agent_info.max_waves_per_cu / agent_info.simds_per_cu);
+  // checkAndSetInfo("(HSA) totalSGPRPerEU", ret.totalSGPRPerEU, ???);
+  // checkAndSetInfo("(HSA) totalVGPRPerEU", ret.totalVGPRPerEU, ???);
```
The point of this PR was to get all the info from a reliable source. If HSA doesn't have all the relevant info, why not just expose the LLVM internal info and use it?

pabloantoniom committed:
Address comments and try to fix CI

pabloantoniom committed:
More fixes and more camelCase

pabloantoniom committed:
Fix build issue after rebase

dhernandez0 commented on mlir/lib/Dialect/Rock/IR/AmdArchDb.cpp:
I thought we were going to change cdna35Info -> cdna4Info, why did you revert that change?

pabloantoniom replied:
Ah, I didn't have that background, change reverted

dhernandez0 commented on mlir/lib/Dialect/Rock/IR/AmdArchDb.cpp:
```diff
@@ -77,24 +85,224 @@ static constexpr AmdArchInfo
+  return ret;
+}
+
+static hsa_status_t AcquireAgentInfo(hsa_agent_t agent, void *data) {
```
why not pass AgentInfo * instead of void*?

dhernandez0 replied:
also, names of functions don't start with capital letters

pabloantoniom replied:
Added a comment on top of the function to explain why it's void*

dhernandez0 commented on mlir/lib/Dialect/Rock/IR/AmdArchDb.cpp:
```diff
@@ -77,24 +85,224 @@ static constexpr AmdArchInfo
+      RET_IF_HSA_ERR(err);
+    }
+  } else {
+    agent_i->numCpus++;
```
looks like numCpus could be a local variable of this function, do we need it in AgentInfo?

pabloantoniom replied:
Hope this is more clear with the comment that I added

dhernandez0 commented on mlir/lib/Dialect/Rock/IR/AmdArchDb.cpp:
```diff
@@ -77,24 +85,224 @@ static constexpr AmdArchInfo
+  // -
+  // https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.cpp
+
+  if (prop->warpSize == 32) {
```
I guess there's no flag for WGP mode?

pabloantoniom replied:
I searched in HIP and HSA but I couldn't find it. Umang said it's actually a compile time thing so I'm not sure how to check it otherwise

dhernandez0 commented on mlir/lib/Dialect/Rock/IR/AmdArchDb.cpp:
```diff
@@ -170,74 +169,61 @@ AmdArchInfo fetchNativeArchInfo(const hipDeviceProp_t &prop) {
+  checkAndSetInfo("(HSA) numEUPerCU", ret.numEUPerCU, agent_info.simds_per_cu);
+  checkAndSetInfo("(HSA) maxWavesPerEU", ret.maxWavesPerEU,
+                  agent_info.max_waves_per_cu / agent_info.simds_per_cu);
+  // checkAndSetInfo("(HSA) totalSGPRPerEU", ret.totalSGPRPerEU, ???);
```
can we add a TODO explaining what params are missing instead of commeted code?

dhernandez0 commented on mlir/lib/Dialect/Rock/IR/AmdArchDb.cpp:
```diff
@@ -170,74 +169,61 @@ AmdArchInfo fetchNativeArchInfo(const hipDeviceProp_t &prop) {
-    checkAndSetInfo("(LLVM) totalVGPRPerEU", ret.totalVGPRPerEU,
-                    st.getTotalNumVGPRs());
-    checkAndSetInfo("(LLVM) waveSize", ret.waveSize, st.getWavefrontSize());
-    checkAndSetInfo("(LLVM) totalSharedMemPerCU", ret.totalSharedMemPerCU,
```
we are missing this one as well

pabloantoniom replied:
In HIP it's called maxSharedMemoryPerMultiProcessor

pabloantoniom committed:
Bring back direct_to_lds_32b to CDNA3

pabloantoniom committed:
Address comments

pabloantoniom committed:
nit: Add suggestion as TODO

pabloantoniom committed:
nit: Add more doc

dhernandez0 commented on mlir/lib/Dialect/Rock/IR/AmdArchDb.cpp:
```diff
@@ -77,32 +86,243 @@ static constexpr AmdArchInfo
+    rdna3Info(GemmFeatures::dot | GemmFeatures::atomic_add |
                   GemmFeatures::atomic_fmax_f32 | GemmFeatures::wmma,
-              /*waveSize=*/32, /*maxWavesPerEU*/ 20, /*totalSGPRPerEU*/ 512,
+              /*waveSize=*/32, /*maxWavesPerEU*/ 16, /*totalSGPRPerEU*/ 800,
```
nit: remove space

pabloantoniom replied:
Why is this not handled by clang-format?

dhernandez0 replied:
I don't know, let's open a ticket

pabloantoniom replied:
I have opened the ticket: https://github.com/ROCm/rocMLIR-internal/issues/2011

It turns out that this wrongly formatted issue happens in a lot of cases in the file, not only with `waveSize`, so I prefer to leave this as it is for now to avoid polluting the PR

dhernandez0 commented on mlir/lib/Dialect/Rock/IR/AmdArchDb.cpp:
```diff
@@ -33,27 +42,27 @@ static constexpr AmdArchInfo
                   GemmFeatures::atomic_add | GemmFeatures::atomic_add_f16 |
                   GemmFeatures::direct_to_lds_32b,
-              /*waveSize=*/64, /*maxWavesPerEU*/ 10, /*totalSGPRPerEU*/ 512,
+              /*waveSize=*/64, /*maxWavesPerEU*/ 8, /*totalSGPRPerEU*/ 800,
```
nit: remove spaces

dhernandez0 commented on mlir/lib/Dialect/Rock/IR/AmdArchDb.cpp:
```diff
@@ -33,27 +42,27 @@ static constexpr AmdArchInfo
                  GemmFeatures::atomic_add_f16,
-             /*waveSize=*/64, /*maxWavesPerEU*/ 8, /*totalSGPRPerEU*/ 512,
-             /*totalVGPRPerEU*/ 512, /*totalSharedMemPerCU*/ 65536,
+             /*waveSize=*/64, /*maxWavesPerEU*/ 10, /*totalSGPRPerEU*/ 800,
```
nit: same here, remove spaces

dhernandez0 commented on mlir/unittests/Dialect/Rock/AmdArchDbTests.cpp:
given that the following params are hardcoded (totalSGPRPerEU, totalVGPRPerEU, defaultFeatures, hasOcpFp8ConversionInstrs) we could have a test for gfx950 and/or navi4 at least? where we hardcode the expected values.

pabloantoniom replied:
Not sure if I understand, you mean something like this?

```
checkAndSetInfo("totalSGPRPerEU", ret.totalSGPRPerEU, 800);
```

dhernandez0 replied:
yes, something like that

pabloantoniom committed:
Fix include (build fails)

mirza-halilcevic commented on mlir/lib/Dialect/Rock/utility/CMakeLists.txt:
Since AmdArchDb is no longer part of MLIRRockUtility, we can revert the changes to this file.

pabloantoniom committed:
Undo changes since AmdArchDb is no longer part of MLIRRockUtility

pabloantoniom committed:
Make HSA optional on Windows; this should fix the UAI build failure

apwojcik commented on mlir/lib/Dialect/Rock/IR/AmdArchDb.cpp:
```diff
@@ -8,12 +8,25 @@
 #include "llvm/Support/ErrorHandling.h"
+#include "llvm/Support/TargetSelect.h"
+
+#include "hip/hip_runtime_api.h"
```
This #include will not compile on Windows. We use MSVC (Visual Studio C++ compiler), which is unaware of HIP and device code. It should be limited to Linux for now.

apwojcik commented on mlir/lib/Dialect/Rock/IR/AmdArchDb.cpp:
```diff
@@ -77,32 +90,252 @@ static constexpr AmdArchInfo
+                          << deviceId << "...\n");
+
+  hipDeviceProp_t prop;
+  if (auto err = hipGetDeviceProperties(&prop, deviceId); err != hipSuccess) {
```
Same here, the MSVC will not compile that code.

apwojcik commented on mlir/lib/Dialect/Rock/IR/CMakeLists.txt:
```diff
@@ -46,3 +46,21 @@ target_link_libraries(MLIRRockOps
   )
+
+# From: https://github.com/ROCm/rocMLIR/blob/b6c726a324d2807494f770ed915af0e54cc6cb3f/external/llvm-project/mlir/lib/ExecutionEngine/CMakeLists.txt#L414
+find_package(hip REQUIRED PATHS ${ROCM_PATH})
```
Please make it Linux only. Due to the compiler we use, the code will not compile.

apwojcik commented on mlir/lib/Dialect/Rock/IR/CMakeLists.txt:
```diff
@@ -46,3 +46,21 @@ target_link_libraries(MLIRRockOps
+find_package(hip REQUIRED PATHS ${ROCM_PATH})
+# We do need both of these. One is for linking, to get the HIP
+# functions.  The other is for compiling, to get -D__HIP_PLATFORM_AMD__
+target_link_libraries(MLIRRockOps PUBLIC
```
Same here

apwojcik commented on mlir/lib/Dialect/Rock/IR/CMakeLists.txt:
```diff
@@ -46,3 +46,21 @@ target_link_libraries(MLIRRockOps
+  hip::host hip::amdhip64
+)
+target_link_libraries(obj.MLIRRockOps PUBLIC
+  hip::host hip::amdhip64
```
and here

apwojcik commented on mlir/lib/Dialect/Rock/utility/Bindings/AmdArchDbBindings.cpp:
```diff
@@ -0,0 +1,53 @@
+//
+//===----------------------------------------------------------------------===//
+
+#include <pybind11/pybind11.h>
```
We do not have pybind11 as a dependency on Windows. This will not compile either.

apwojcik commented on mlir/lib/Dialect/Rock/utility/Bindings/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+find_package(Python3 COMPONENTS Interpreter Development REQUIRED)
```
Same here; we do not require pybind11 as a dependency on Windows. This will not compile.

apwojcik commented on mlir/lib/Dialect/Rock/IR/AmdArchDb.cpp:
```diff
@@ -77,32 +90,252 @@ static constexpr AmdArchInfo
+  uint32_t maxWavesPerCU;
+};
+
+AmdArchInfo fetchNativeArchInfo(const hipDeviceProp_t &prop,
```
Can we move that to the migraphx_gpu target instead? To read the required properties there and pass them to the rocMLIR instead of makeing HIP a dependency of rocMLIR?

Currently the HIP dependency compilcates a lot on Windows for rocMLIR.

pabloantoniom committed:
Do not build native arch under Windows since our CI does not support HIP/HSA

dhernandez0 commented on mlir/lib/Dialect/Rock/utility/Bindings/CMakeLists.txt:
```diff
@@ -0,0 +1,11 @@
+
+    pybind11_add_module(amd_arch_db AmdArchDbBindings.cpp)
+    target_link_libraries(amd_arch_db PUBLIC MLIRRockUtility)
+endif()
```
new line

dhernandez0 commented on mlir/unittests/Dialect/Rock/CMakeLists.txt:
```diff
@@ -1,9 +1,17 @@
   MLIRRockOps
   MLIRRockUtility
-)
+)
```
new line

pabloantoniom committed:
Try to use this hack to fix Windows CI

pabloantoniom committed:
Add newlines

pabloantoniom committed:
Ignore problematic files for clang-tidy
